tensordict/dynamic_shaped_tensor.py

```diff
@@ -0,0 +1,74 @@
+import numpy as np
+import torch
+
+from tensordict import MemmapTensor
+
+
+class DynamicShapeTensor:
+    def __init__(
+        self,
+        shape: torch.Size,  # -1 indicates variable size
+        total_elts: int,  # size of the storage
+        total_index: int,  # number of indexable items along stack dim
+        stack_dim: int = 0,
+    ):
+        if stack_dim != 0:
+            raise NotImplementedError(stack_dim)
+        self.offsets = MemmapTensor(total_index, dtype=torch.int64).as_tensor()
+        self.total_elts = total_elts
+        assert sum(s == -1 for s in shape) == 1
+        assert sum(s > 0 for s in shape) == len(shape) - 1
+        self.shape = shape
+        self.shapes = MemmapTensor(total_index, dtype=torch.int64).as_tensor()
+        self.storage = MemmapTensor(total_elts).as_tensor()
+        self.variable_shape = sum(np.cumprod([s > 0 for s in shape]))
+        self.total_index = total_index
+
+    def __setitem__(self, idx: int, value):
+        assert isinstance(idx, int)
+        assert value.ndim == len(self.shape)
+        assert (s1 == s2 for (s1, s2) in zip(value.shape, self.shape) if s2 > 0)
+        prev_offset = self.offsets[idx]
+        offset = value.numel()
+        if idx < 0:
```

Comment:
I think you can re-use Numpy's indexing behavior without reimplementing it here. This can be achieved by initializing a field with `np.arange` with the number of elements, and apply `idx` to this field, this will give you the indices you need, always positive. This will also make sure that slicing also works

Reply:
I wonder what's the overhead of doing that? What if you have >100M elements for instance?

Reply:
There will be memory overhead, yes. Similar to the offsets array. And since here we don't need to support advanced indexing in getitem, then it does not make sense, it's just unnecessary overhead. Only if in the future we want to for example to be able to pass a multi-dimensional tensor to getitem (in the way it can be passed in numpy), it would make sense